pgrx-examples/operators/src/pgvarlena.rs

```diff
@@ -0,0 +1,135 @@
+/*
+Portions Copyright 2019-2021 ZomboDB, LLC.
+Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+use pgrx::prelude::*;
+use pgrx::StringInfo;
+use serde::{Deserialize, Serialize};
+use std::ffi::CStr;
+use std::fmt::{Display, Formatter, Write};
+
+/// standard Rust equality/comparison derives
+#[derive(Eq, PartialEq, Ord, Hash, PartialOrd)]
+
+/// Support using this struct as a Postgres type, which the easy way requires Serde
+#[derive(PostgresType, Serialize, Deserialize)]
+
+/// automatically generate =, <> SQL operator functions
+#[derive(PostgresEq)]
+
+/// automatically generate <, >, <=, >=, and a "_cmp" SQL functions
+/// When "PostgresEq" is also derived, pgrx also creates an "opclass" (and family)
+/// so that the type can be used in indexes `USING btree`
+#[derive(PostgresOrd)]
+
+/// automatically generate a "_hash" function, and the necessary "opclass" (and family)
+/// so the type can also be used in indexes `USING hash`
+#[derive(PostgresHash)]
+
+/// Necessary derives for generally working with [`PgVarlena`]
+#[derive(Copy, Clone)]
+
+/// For unit tests
+#[derive(Debug)]
+
+/// we want this to look like a C struct
+#[repr(C)]
+
+/// Indicate to `#[derive(PostgresType)]` that we'll write our own input and output function
+#[pgvarlena_inoutfuncs]
+pub struct PgVarlenaThing {
+    a: u64,
+    b: u64,
+    c: i32,
+    d: [u8; 5],
+}
+
+impl PgVarlenaThing {
+    pub fn new(a: u64, b: u64, c: i32, d: [u8; 5]) -> Self {
+        Self { a, b, c, d }
+    }
+}
+
+impl Display for PgVarlenaThing {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.a.to_string())?;
+        f.write_char(';')?;
+        f.write_str(&self.b.to_string())?;
+        f.write_char(';')?;
+        f.write_str(&self.c.to_string())?;
+        f.write_char(';')?;
+        f.write_str(&format!("{:?}", self.d))
+    }
+}
+
+// implement function to parse string representation of this type and a function to convert it
+// to a string
+impl PgVarlenaInOutFuncs for PgVarlenaThing {
+    fn input(input: &CStr) -> PgVarlena<Self>
```

Comment:
This could also be cleaned up, but it's a bit more involved and I'm not sure how much we care.